File: elixir-phoenix/lib/chopchop_web/controllers/link_controller.ex
defmodule ChopchopWeb.LinkController do
  use ChopchopWeb, :controller

  alias Chopchop.{Codegen, Links}

  # GET /health
  def health(conn, _params) do
    json(conn, %{status: "ok", language: "elixir", framework: "phoenix"})
  end

  # POST /chop
  def create(conn, params) do
    url = Map.get(params, "url")
    custom_code = Map.get(params, "custom_code")
    expires_in = Map.get(params, "expires_in")

    with :ok <- validate_url(url),
         {:ok, code} <- resolve_code(custom_code),
         {:ok, expires_at} <- parse_expires_in(expires_in) do
      now = DateTime.utc_now() |> DateTime.truncate(:second)

      case Links.create_link(%{code: code, url: url, created_at: now, expires_at: expires_at}) do
        {:ok, link} ->
          base = "#{conn.scheme}://#{conn.host}#{port_suffix(conn)}"

          conn
          |> put_status(201)
          |> json(%{
            code: link.code,
            short_url: "#{base}/#{link.code}",
            url: link.url,
            created_at: format_datetime(link.created_at),
            expires_at: format_datetime(link.expires_at)
          })

        {:error, _changeset} ->
          conn |> put_status(500) |> json(%{error: "Failed to create link"})
      end
    else
      {:error, :invalid_url} ->
        conn |> put_status(400) |> json(%{error: "Invalid or missing URL"})

      {:error, :invalid_custom_code} ->
        conn
        |> put_status(400)
        |> json(%{error: "custom_code must be 3–20 alphanumeric characters or hyphens"})

      {:error, :code_taken} ->
        conn |> put_status(409) |> json(%{error: "Custom code already taken"})

      {:error, :invalid_expires_in} ->
        conn
        |> put_status(400)
        |> json(%{error: "expires_in must be a positive integer no greater than 2592000"})

      {:error, :too_many_collisions} ->
        conn |> put_status(500) |> json(%{error: "Failed to generate a unique code"})
    end
  end

  # GET /stats/:code
  def stats(conn, %{"code" => code}) do
    case Links.get_stats(code) do
      nil ->
        conn |> put_status(404) |> json(%{error: "Link not found"})

      {link, total, recent} ->
        json(conn, %{
          code: link.code,
          url: link.url,
          created_at: format_datetime(link.created_at),
          expires_at: format_datetime(link.expires_at),
          total_clicks: total,
          recent_clicks:
            Enum.map(recent, fn c ->
              %{
                clicked_at: format_datetime(c.clicked_at),
                referer: c.referer,
                user_agent: c.user_agent
              }
            end)
        })
    end
  end

  # GET /:code
  def resolve(conn, %{"code" => code}) do
    case Links.get_by_code(code) do
      nil ->
        conn |> put_status(404) |> json(%{error: "Link not found"})

      link ->
        if expired?(link) do
          conn |> put_status(410) |> json(%{error: "Link has expired"})
        else
          ip = conn.remote_ip |> ip_to_string()
          user_agent = get_req_header(conn, "user-agent") |> List.first()
          referer = get_req_header(conn, "referer") |> List.first()

          Links.record_click(link, ip, user_agent, referer)

          conn
          |> put_resp_header("location", link.url)
          |> send_resp(301, "")
        end
    end
  end

  # ── helpers ────────────────────────────────────────────────────────────────

  defp validate_url(url) when is_binary(url) do
    case URI.parse(url) do
      %URI{scheme: scheme, host: host} when scheme in ["http", "https"] and is_binary(host) ->
        if String.contains?(host, "."), do: :ok, else: {:error, :invalid_url}

      _ ->
        {:error, :invalid_url}
    end
  end

  defp validate_url(_), do: {:error, :invalid_url}

  defp resolve_code(nil) do
    Codegen.generate()
  end

  defp resolve_code(code) when is_binary(code) do
    if Regex.match?(~r/^[a-zA-Z0-9\-]{3,20}$/, code) do
      if Links.code_taken?(code) do
        {:error, :code_taken}
      else
        {:ok, code}
      end
    else
      {:error, :invalid_custom_code}
    end
  end

  defp resolve_code(_), do: {:error, :invalid_custom_code}

  defp parse_expires_in(nil), do: {:ok, nil}

  defp parse_expires_in(seconds) when is_integer(seconds) and seconds > 0 and seconds <= 2_592_000 do
    {:ok, DateTime.utc_now() |> DateTime.add(seconds, :second) |> DateTime.truncate(:second)}
  end

  defp parse_expires_in(_), do: {:error, :invalid_expires_in}

  defp expired?(%{expires_at: nil}), do: false

  defp expired?(%{expires_at: expires_at}) do
    DateTime.compare(DateTime.utc_now(), expires_at) == :gt
  end

  defp format_datetime(nil), do: nil

  defp format_datetime(%DateTime{} = dt) do
    dt |> DateTime.truncate(:second) |> DateTime.to_iso8601() |> String.replace("Z", "+00:00")
  end

  defp ip_to_string({a, b, c, d}), do: "#{a}.#{b}.#{c}.#{d}"

  defp ip_to_string({a, b, c, d, e, f, g, h}) do
    [a, b, c, d, e, f, g, h]
    |> Enum.map(&Integer.to_string(&1, 16))
    |> Enum.join(":")
  end

  defp port_suffix(%Plug.Conn{scheme: :http, port: 80}), do: ""
  defp port_suffix(%Plug.Conn{scheme: :https, port: 443}), do: ""
  defp port_suffix(%Plug.Conn{port: port}), do: ":#{port}"
end

File: java-springboot/src/main/java/com/chopchop/controller/LinkController.java
package com.chopchop.controller;

import com.chopchop.model.Click;
import com.chopchop.model.Link;
import com.chopchop.repository.ClickRepository;
import com.chopchop.repository.LinkRepository;
import com.chopchop.service.CodeGenerator;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

@RestController
public class LinkController {

    private static final Pattern CODE_PATTERN = Pattern.compile("^[a-zA-Z0-9\\-]{3,20}$");
    private static final int MAX_EXPIRES_IN = 2_592_000;
    private static final DateTimeFormatter ISO = DateTimeFormatter.ISO_OFFSET_DATE_TIME;

    private final LinkRepository linkRepository;
    private final ClickRepository clickRepository;
    private final CodeGenerator codeGenerator;

    public LinkController(LinkRepository linkRepository, ClickRepository clickRepository, CodeGenerator codeGenerator) {
        this.linkRepository = linkRepository;
        this.clickRepository = clickRepository;
        this.codeGenerator = codeGenerator;
    }

    // GET /health
    @GetMapping("/health")
    public Map<String, String> health() {
        return Map.of("status", "ok", "language", "java", "framework", "spring boot");
    }

    // POST /chop
    @PostMapping("/chop")
    public ResponseEntity<Object> chop(@RequestBody Map<String, Object> body, HttpServletRequest request) {
        String url = asString(body.get("url"));
        Object customCodeObj = body.get("custom_code");
        Object expiresInObj = body.get("expires_in");

        if (url == null || !isValidUrl(url)) {
            return error(HttpStatus.BAD_REQUEST, "Invalid or missing URL");
        }

        String code;
        if (customCodeObj != null) {
            String customCode = asString(customCodeObj);
            if (customCode == null || !CODE_PATTERN.matcher(customCode).matches()) {
                return error(HttpStatus.BAD_REQUEST, "custom_code must be 3–20 alphanumeric characters or hyphens");
            }
            if (linkRepository.existsByCode(customCode)) {
                return error(HttpStatus.CONFLICT, "Custom code already taken");
            }
            code = customCode;
        } else {
            code = codeGenerator.generate();
        }

        OffsetDateTime expiresAt = null;
        if (expiresInObj != null) {
            if (!(expiresInObj instanceof Integer expiresIn) || expiresIn <= 0 || expiresIn > MAX_EXPIRES_IN) {
                return error(HttpStatus.BAD_REQUEST, "expires_in must be a positive integer no greater than 2592000");
            }
            expiresAt = OffsetDateTime.now(ZoneOffset.UTC).plusSeconds(expiresIn);
        }

        Link link = new Link();
        link.setCode(code);
        link.setUrl(url);
        link.setCreatedAt(OffsetDateTime.now(ZoneOffset.UTC));
        link.setExpiresAt(expiresAt);
        linkRepository.save(link);

        String base = request.getScheme() + "://" + request.getServerName()
                + (request.getServerPort() != 80 && request.getServerPort() != 443
                        ? ":" + request.getServerPort() : "");

        Map<String, Object> resp = new LinkedHashMap<>();
        resp.put("code", link.getCode());
        resp.put("short_url", base + "/" + link.getCode());
        resp.put("url", link.getUrl());
        resp.put("created_at", link.getCreatedAt().format(ISO));
        resp.put("expires_at", link.getExpiresAt() != null ? link.getExpiresAt().format(ISO) : null);

        return ResponseEntity.status(HttpStatus.CREATED).body(resp);
    }

    // GET /stats/:code
    @GetMapping("/stats/{code}")
    public ResponseEntity<Object> stats(@PathVariable String code) {
        Link link = linkRepository.findByCode(code).orElse(null);
        if (link == null) {
            return error(HttpStatus.NOT_FOUND, "Link not found");
        }

        long totalClicks = clickRepository.countByLinkId(link.getId());
        List<Click> recent = clickRepository.findTop10ByLinkIdOrderByClickedAtDesc(link.getId());

        Map<String, Object> resp = new LinkedHashMap<>();
        resp.put("code", link.getCode());
        resp.put("url", link.getUrl());
        resp.put("created_at", link.getCreatedAt().format(ISO));
        resp.put("expires_at", link.getExpiresAt() != null ? link.getExpiresAt().format(ISO) : null);
        resp.put("total_clicks", totalClicks);
        resp.put("recent_clicks", recent.stream().map(c -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("clicked_at", c.getClickedAt().format(ISO));
            m.put("referer", c.getReferer());
            m.put("user_agent", c.getUserAgent());
            return m;
        }).toList());

        return ResponseEntity.ok(resp);
    }

    // GET /:code
    @GetMapping("/{code}")
    public ResponseEntity<Object> redirect(@PathVariable String code, HttpServletRequest request) {
        Link link = linkRepository.findByCode(code).orElse(null);
        if (link == null) {
            return error(HttpStatus.NOT_FOUND, "Link not found");
        }

        if (link.getExpiresAt() != null && link.getExpiresAt().isBefore(OffsetDateTime.now(ZoneOffset.UTC))) {
            return error(HttpStatus.GONE, "Link has expired");
        }

        Click click = new Click();
        click.setLink(link);
        click.setClickedAt(OffsetDateTime.now(ZoneOffset.UTC));
        click.setIpAddress(clientIp(request));
        click.setUserAgent(request.getHeader("User-Agent"));
        click.setReferer(request.getHeader("Referer"));
        clickRepository.save(click);

        return ResponseEntity.status(HttpStatus.MOVED_PERMANENTLY)
                .location(URI.create(link.getUrl()))
                .build();
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static boolean isValidUrl(String url) {
        if (!url.startsWith("http://") && !url.startsWith("https://")) return false;
        try {
            URI uri = URI.create(url);
            String host = uri.getHost();
            return host != null && host.contains(".");
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    private static String asString(Object o) {
        return (o instanceof String s) ? s : null;
    }

    private static String clientIp(HttpServletRequest request) {
        String forwarded = request.getHeader("X-Forwarded-For");
        if (forwarded != null && !forwarded.isBlank()) {
            return forwarded.split(",")[0].trim();
        }
        return request.getRemoteAddr();
    }

    private static ResponseEntity<Object> error(HttpStatus status, String message) {
        return ResponseEntity.status(status).body(Map.of("error", message));
    }
}
